### PR TITLE
hiop: Fix how ReSolve option is passed

### PR DESCRIPTION
**Pull request by cameronrutherford**
Was getting some cryptic errors building hiop~cuda and it looked like some extra whitespace was being added to the CMake line due to a stray CLI argument from an undefined `args.extend` input.

This should fix that issue - see error [pipeline](https://github.com/LLNL/hiop/actions/runs/10741214413/job/29791072221)

### CODE REVIEW

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 1]
```diff
@@ -189,6 +189,9 @@ def cmake_args(self):
                 ]
             )
 
+        with when("+cuda @:0.7.1"):
+            args.extend(self.define_from_variant("HIOP_USE_RESOLVE", "cusolver_lu"))
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py]: Maybe there is a cleaner way to do this... It's functional and I am unsure if this was even necessary to resolve the bug we were seeing in the GitHub pipelines, but seems like a sensible change to remove additional whitespace in CMake line.

**Comment by cameronrutherford** [comment]
The HiOp pipelines are now passing.

I noticed that there was a random whitespace in the CMake command before I finally figured out that I didn't clone the HiOp repo in the pipeline, and that's actually why CMake was failing.

Makes me wonder if spack should auto-remove whitespace to cover edge cases like this, but I suppose CMake might just also ignore whitespace like this.

**Comment by cameronrutherford** [comment]
I don't know why CI is failing (test timed out?).

This change is sensible for builds without cuda. Cc @nychiang

**Comment by nychiang** [comment]
@cameronrutherford 
it seems to me the argument `-DHIOP_USE_RESOLVE` is incorrectly defined.
The log shows the followings:
'''
'-' 'D' 'H' 'I' 'O' 'P' '_' 'U' 'S' 'E' '_' 'R' 'E' 'S' 'O' 'L' 'V' 'E' ':' 'B' 'O' 'O' 'L' '=' 'O' 'F' 'F' '-DHIOP_BUILD_STATIC:BOOL=ON' '-DBLAS_FOUND:BOOL=ON'
'''
Any thought about why this happens?

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 2]
```diff
@@ -189,6 +189,9 @@ def cmake_args(self):
                 ]
             )
 
+        with when("+cuda @:0.7.1"):
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py]:
```suggestion
        with when("+cuda @0.7.1:"):
```

**Comment by cameronrutherford** [comment]
> @cameronrutherford it seems to me the argument `-DHIOP_USE_RESOLVE` is incorrectly defined. The log shows the followings: ''' '-' 'D' 'H' 'I' 'O' 'P' '_' 'U' 'S' 'E' '_' 'R' 'E' 'S' 'O' 'L' 'V' 'E' ':' 'B' 'O' 'O' 'L' '=' 'O' 'F' 'F' '-DHIOP_BUILD_STATIC:BOOL=ON' '-DBLAS_FOUND:BOOL=ON' ''' Any thought about why this happens?

Not really sure... I noticed an issue with the syntax when I looked at the code again, but I think I am doing everything right. We shall see what CI says after this...

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 3]
```diff
@@ -189,6 +189,9 @@ def cmake_args(self):
                 ]
             )
 
+        with when("+cuda @0.7.1:"):
+            args.extend(self.define_from_variant("HIOP_USE_RESOLVE", "cusolver_lu"))
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py]:
```suggestion
            args.extend([self.define_from_variant("HIOP_USE_RESOLVE", "cusolver_lu")])
```
Maybe this fixes it

**Comment by cameronrutherford** [comment]
@nychiang I am closing this, but I think it should be merged. CI passed on the last pipeline. Not sure why we never merged.

**Comment by pelesh** [comment]
I think this should be merged, too. Reopening.

**Review by pelesh — APPROVED** [review]
Looks good.

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 4]
```diff
@@ -203,6 +203,9 @@ def cmake_args(self):
                 ]
             )
 
+        with when("+cuda @0.7.1:"):
+            args.extend([self.define_from_variant("HIOP_USE_RESOLVE", "cusolver_lu")])
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hiop/package.py]: This is wrong. The context manager:
```
with when(...):
    ...
```
works at class level, and only for directives. What was this meant to do?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 4 above)
Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/hiop/package.py]:
```suggestion
        if self.spec.satisfies("+cuda @0.7.1:"):
            args.extend([self.define_from_variant("HIOP_USE_RESOLVE", "cusolver_lu")])
```

This is what we could change it to?

The idea is that `HIOP_USE_RESOLVE` is only an option in `@0.7.1:`, and then I think the `+cuda` requirement is redundant. We want to `define_from_variant` based on `cusolver_lu`.

Now that I think about it, won't CMake just reference this as an unused variable in earlier versions? I assume that we can get away with lumping this into our other `args.extend` call if so.

Comment (alalazo) [reply var/spack/repos/builtin/packages/hiop/package.py]: > Now that I think about it, won't CMake just reference this as an unused variable in earlier versions?

Yes, it should

Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/hiop/package.py]: Okay I simplified and moved this as discussed.

**Comment by cameronrutherford** [comment]
Now it looks like this PR is basically a no-op after making adjustments based on review? @pelesh or @nychiang I've forgotten the context behind this PR, and so perhaps it isn't necessary anymore.

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 4 above)
Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/hiop/package.py]: Changing back turned this into a no-op of a PR... I think the issue is that when building `~cuda`, the option got mangled and CMake threw a fit. Unless a maintainer of hiop can remember why / reproduce the issue, or if there are build failures that need addressing, then I think we can close the PR.

**Comment by pelesh** [comment]
> Now it looks like this PR is basically a no-op after making adjustments based on review? @pelesh or @nychiang I've forgotten the context behind this PR, and so perhaps it isn't necessary anymore.

It seems we cannot reproduce the issue addressed in this PR anymore. I suggest we close.
